Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/ChangeEvent.swift

```diff
@@ -0,0 +1,168 @@
+     */
+    public let fullDocument: DocumentT?
+    /// The namespace (database and or collection) affected by the event.
+    public let ns: MongoNamespace
+    /**
+     The ObjectID of the document created or modified by the
+     insert, replace, delete, update operations (i.e. CRUD operations).
+     */
+    public let documentKey: Document
+    /**
+     A document describing the fields that were updated or removed by
+     the update operation.
+     */
+    public let updateDescription: UpdateDescription?
+    /// Whether or not this ChangeEvent has pending writes.
+    public let hasUncommittedWrites: Bool
+
+    public static func == (lhs: ChangeEvent<DocumentT>, rhs: ChangeEvent<DocumentT>) -> Bool {
+        return bsonEquals(lhs.documentKey, rhs.documentKey)
+    }
+
+    public func hash(into hasher: inout Hasher) {
+        hasher.combine(documentKey.canonicalExtendedJSON)
+    }
+
+    /**
+     Transform a ChangeEvent of type Document to type T.
+
+     - parameter changeEvent: the ChangeEvent to transform
+     - returns: a ChangeEvent where the fullDocument is of type T
+     */
+    static func transform<T: Codable>(changeEvent: ChangeEvent<Document>) -> ChangeEvent<T> {
+        var fullDocument: T?
+        if let actualFullDocument = changeEvent.fullDocument {
+            fullDocument = try! BSONDecoder().decode(T.self,
+                                                     from: actualFullDocument)
+        }
+        return ChangeEvent<T>.init(
+            id: changeEvent.id,
+            operationType: changeEvent.operationType,
+            fullDocument: fullDocument,
+            ns: changeEvent.ns,
+            documentKey: changeEvent.documentKey,
+            updateDescription: changeEvent.updateDescription,
+            hasUncommittedWrites: changeEvent.hasUncommittedWrites)
+    }
+
+    /**
+     Generates a change event for a local insert of the given document in the given namespace.
+
+     - parameter namespace: the namespace where the document was inserted.
+     - parameter document: the document that was inserted.
+     - returns a change event for a local insert of the given document in the given namespace.
+     */
+    static func changeEventForLocalInsert(namespace: MongoNamespace,
+                                          document: Document,
+                                          writePending: Bool) -> ChangeEvent<Document> {
+        let docId = document["_id"]
+        return ChangeEvent<Document>.init(
+            id: Document(),
+            operationType: .insert,
+            fullDocument: document,
+            ns: namespace,
+            documentKey: ["_id": docId],
+            updateDescription: nil,
+            hasUncommittedWrites: writePending)
+    }
+
+    /**
+     Generates a change event for a local update of a document in the given namespace referring
+     to the given document _id.
+
+     - parameter namespace: the namespace where the document was inserted.
```

Comment:
[nit] updated

Reply:
+1, should be [typo] and should be fixed